File: python/2024/01_historian_hysteria/aoc202401.py
"""AoC 1, 2024: Historian Hysteria."""

# Standard library imports
import collections
import pathlib
import sys


def parse_data(puzzle_input):
    """Parse input."""
    return list(
        zip(
            *[
                [int(number) for number in line.split()]
                for line in puzzle_input.split("\n")
            ]
        )
    )


def part1(data):
    """Solve part 1."""
    left, right = data
    return sum(
        abs(first - second) for first, second in zip(sorted(left), sorted(right))
    )


def part2(data):
    """Solve part 2."""
    left, right = data
    counts = collections.Counter(right)
    return sum(number * counts.get(number, 0) for number in left)


def solve(puzzle_input):
    """Solve the puzzle for the given input."""
    data = parse_data(puzzle_input)
    yield part1(data)
    yield part2(data)


if __name__ == "__main__":
    for path in sys.argv[1:]:
        print(f"\n{path}:")
        solutions = solve(puzzle_input=pathlib.Path(path).read_text().rstrip())
        print("\n".join(str(solution) for solution in solutions))

File: elixir/lib/2024/01_historian_hysteria/aoc202401.ex
defmodule AOC2024.Day01 do
  @moduledoc """
  Advent of Code 2024, day 1: Historian Hysteria.
  """
  require AOC

  @doc """
  Parse input.
  """
  def parse(puzzle_input) do
    puzzle_input
    |> String.split("\n", trim: true)
    |> Enum.map(fn line ->
      line
      |> String.split()
      |> Enum.map(&String.to_integer/1)
      |> then(fn [left, right] -> {left, right} end)
    end)
    |> Enum.unzip()
  end

  @doc """
  Solve part 1.
  """
  def part1({left, right}) do
    Enum.zip(Enum.sort(left), Enum.sort(right))
    |> Enum.map(fn {first, second} -> abs(second - first) end)
    |> Enum.sum()
  end

  @doc """
  Solve part 2.
  """
  def part2({left, right}) do
    counts = Enum.frequencies(right)

    left
    |> Enum.map(fn number -> number * Map.get(counts, number, 0) end)
    |> Enum.sum()
  end

  def main(args) do
    Enum.map(args, fn path -> AOC.solve(path, &parse/1, &part1/1, &part2/1) end)
  end
end
